org.eclipse.lsp4j/src/main/java/org/eclipse/lsp4j/services/TextDocumentService.java

```diff
@@ -374,5 +376,16 @@ default void willSave(WillSaveTextDocumentParams params) {}
 	default CompletableFuture<List<FoldingRange>> foldingRange(FoldingRangeRequestParams params) {
 		throw new UnsupportedOperationException();
 	}
+
+	/**
+	 * The prepare rename request is sent from the client to the server to setup and test the validity of a rename
+	 * operation at a given location.
+	 * 
+	 * Since version 3.12.0
+	 */
+	@JsonRequest
+	default CompletableFuture<Either<Range, PrepareRenameResult>> prepareRename(TextDocumentPositionParams params) {
```

Comment:
Parsing this requires a TypeAdapterFactory registered with a `@ResponseJsonAdapter` annotation (see `documentSymbol` and `codeAction`).